### ISSUE TITLE
d3.transition() crash in IE<11

### ISSUE THREAD

**Issue by bensiron** [mbostock/d3#2174]
When using D3 with C3.js in IE10 or lower, C3 makes a call to d3.transition() without any arguments, and this crashes in IE10 and lower because d3_selectionRoot has no transition method defined.  This in turn is because d3_subclass fakes inheritance in IE<11 where there is no __proto__ property by doing:
for (var property in prototype) object[property] = prototype[property];  at the time that d3_selectionRoot is created.  Unfortunately, at the time d3_selectionRoot is created, d3_selectionPrototype does not yet have the transition method defined.  If you move   d3_selectionRoot = d3.select(d3_documentElement);
to after all the d3_selectionPrototype stuff is setup at line 8573, the problem is solved. 

Is this a real issue, or am I or C3.js simply using D3 wrongly?

**Comment by mbostock** [comment]
This is a bug in D3, but it’s also an inherent weakness in IE10-.

The problem is IE10 and below don’t support setting \_\_proto\_\_, so there’s no good way to extend the built-in array type in those browsers; instead, we have to assign the methods manually on each selection. The bug results in an ordering problem in the way the imports are resolved, such that the `d3_selectionRoot` instance is created before the `d3_selectionPrototype` is fully defined, and so this selection instance doesn’t have all the intended methods available.

The simplest way to fix this is to change the order of the imports to defer the instantiation of `d3_selectionRoot` until after the `d3_selectionPrototype` is fully defined. (I believe there’s already been at least one instance of this bug appearing the past, and a similar fix being applied.)

A more ambitious way of fixing this might be to abandon the approach of extending the built-in array type, which will always be a hack we can exclusively support ES6. This would also fix #1805, but it would not be considered a backwards-compatible change since there is undoubtedly some code in the wild that depends on selection instances being arrays.

**Comment by wfortin** [comment]
I'm experiencing the same problem with c3 on IE9 and IE10. I have found that using the latest d3 (3.5.2) instead of the same version as c3's examples causes the error.

Make sure you use the same versions of d3 and c3 as in the c3 examples. The current version of c3 is 0.4.8 and they depend on d3 version 3.4.3. Downgrading to 3.4.3 fixed the issue.

All releases of d3 are available here: https://github.com/mbostock/d3/releases

**Comment by wfortin** [comment]
I've added an issue to C3 to document (or provide) the officially supported d3 version. https://github.com/masayuki0812/c3/issues/893

Hope this helps.

**Comment by mbostock** [comment]
Duplicate of #1491 which was fixed by #2225 and will be available in 3.5.4.
